### PR TITLE
solution for jv-array-list

### PR DESCRIPTION
**Pull request by vladis9**

### CODE REVIEW

File: src/main/java/core/basesyntax/ArrayList.java [hunk 1]
```diff
@@ -1,48 +1,102 @@
...
+        size--;
+        return removed;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        for (int i = 0; i < size; i++) {
+            if (elements[i] == element || elements[i] != null && elements[i].equals(element)) {
+                return remove(i);
```
Comment (overpathz) [inline src/main/java/core/basesyntax/ArrayList.java:70]: Everything is okay, but for improvement, we can create separate method for searching index for an element.

File: src/main/java/core/basesyntax/ArrayList.java [hunk 2]
```diff
@@ -1,48 +1,102 @@
...
+    public ArrayList() {
+        elements = (T[]) new Object[DEFAULT_CAPACITY];
+    }
+
     @Override
     public void add(T value) {
-
+        if (size == elements.length) {
+            grow();
+        }
```
Comment (overpathz) [inline src/main/java/core/basesyntax/ArrayList.java:19]: you can move it to separate method(for example, ensureCapacity()) and use it in add(T value, int index).

File: src/main/java/core/basesyntax/ArrayList.java [hunk 3]
```diff
@@ -1,48 +1,102 @@
...
+            add(value);
+            return;
+        }
+        checkIndex(index);
+        if (size + 1 == elements.length) {
+            grow();
+        }
+        System.arraycopy(elements, index, elements, index + 1, size - index);
+        elements[index] = value;
+        size++;
```
Comment (overpathz) [inline src/main/java/core/basesyntax/ArrayList.java:35]:
```suggestion
        checkIndexForAdd(index);
        ensureCapacity();
        System.arraycopy(elementData, index, elementData, index + 1, size - index);
        elementData[index] = value;
        size++;
```

**Review by overpathz — CHANGES_REQUESTED** [review]
Nice job.
Refactor some remarks

File: src/main/java/core/basesyntax/ArrayList.java [hunk 4]
```diff
@@ -1,48 +1,114 @@
...
+        size--;
+        return removed;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        int index = getIndex(element);
+        if (index == -1) {
+            throw new NoSuchElementException("There is no element such element");
```
Comment (okuzan) [inline src/main/java/core/basesyntax/ArrayList.java:62]: add element to the message

File: src/main/java/core/basesyntax/ArrayList.java [hunk 5]
```diff
@@ -1,48 +1,114 @@
...
+        if (index >= size || index < 0) {
+            throwIndexOutOfBounds(index);
+        }
+    }
+
+    private void checkIndexForAdd(int index) {
+        if (index > size || index < 0) {
+            throwIndexOutOfBounds(index);
+        }
+    }
```
Comment (okuzan) [inline src/main/java/core/basesyntax/ArrayList.java:87]: could use the standard checker if placed after ensureCapacity

File: src/main/java/core/basesyntax/ArrayList.java [hunk 6]
```diff
@@ -1,48 +1,114 @@
...
+            if (elements[i] == element || elements[i] != null && elements[i].equals(element)) {
+                return i;
+            }
+        }
+        return -1;
+    }
+
+    private void throwIndexOutOfBounds(int index) {
+        throw new ArrayListIndexOutOfBoundsException("Wrong index" + index + "! "
+                + "Index value must be less than size");
```
Comment (okuzan) [inline src/main/java/core/basesyntax/ArrayList.java:112]: include size in the message

**Review by okuzan — APPROVED** [review]
Well done! Minor remarks for further improvement

**Pull request closed without merging**